via/static/scripts/video_player/components/ToastMessages.tsx

```diff
@@ -0,0 +1,187 @@
+ */
+function ToastMessageItem({ message, onDismiss }: ToastMessageItemProps) {
+  // Capitalize the message type for prepending; Don't prepend a message
+  // type for "notice" messages
+  const prefix =
+    message.type !== 'notice'
+      ? `${message.type.charAt(0).toUpperCase() + message.type.slice(1)}: `
+      : '';
+
+  return (
+    <Callout
+      classes={classnames({
+        'sr-only': message.visuallyHidden,
+      })}
+      status={message.type}
+      onClick={() => onDismiss(message.id)}
+      variant="raised"
+    >
+      <strong>{prefix}</strong>
+      {message.message}
+    </Callout>
+  );
+}
+
+const BaseToastMessageTransition: TransitionComponent = ({
+  direction = 'in',
+  onTransitionEnd,
+  children,
+}) => {
+  const isDismissed = direction === 'out';
+  const containerRef = useRef<HTMLDivElement>(null);
+  const handleAnimation = (e: AnimationEvent) => {
+    // Ignore non-relevant animations:
+    // * Happening on children elements
+    // * Animations which are not relevant for toast messages getting "in" or "out"
+    if (
+      e.target !== containerRef.current ||
+      !['slide-in-from-right', 'fade-in', 'fade-out'].includes(e.animationName)
+    ) {
+      return;
+    }
+
+    onTransitionEnd?.(direction);
+  };
+
+  return (
+    <div
+      onAnimationEnd={handleAnimation}
+      ref={containerRef}
+      className={classnames('relative w-full container', {
+        // Only ever slide-from-right if motion-safe is preferred
+        'lg:motion-safe:animate-slide-in-from-right animate-fade-in':
+          !isDismissed,
+        // Only ever fade-in if motion-reduction is preferred
+        'motion-reduce:animate-fade-in': !isDismissed,
+        'animate-fade-out': isDismissed,
+      })}
+    >
+      {children}
+    </div>
+  );
+};
+
+export type ToastMessagesProps = {
+  messages: ToastMessage[];
+  onMessageDismiss: (id: string) => void;
+  _setTimeout?: typeof setTimeout;
+};
+
+/**
+ * A collection of toast messages. These are rendered within an `aria-live`
+ * region for accessibility with screen readers.
+ */
+export default function ToastMessages({
+  messages,
+  onMessageDismiss,
+  /* istanbul ignore next - test seam */
+  _setTimeout = setTimeout,
+}: ToastMessagesProps) {
+  const [dismissedMessages, setDismissedMessages] = useState<string[]>([]);
+  const scheduledMessages = useRef(new Set<string>());
+
+  const dismissMessage = useCallback(
+    (id: string) => setDismissedMessages(ids => [...ids, id]),
+    []
+  );
+  const scheduleMessageDismiss = useCallback(
+    (id: string) => {
+      if (scheduledMessages.current.has(id)) {
+        return;
+      }
+
+      // Track that this message has been scheduled to be dismissed. After a
+      // period of time, actually dismiss it
+      scheduledMessages.current.add(id);
+      _setTimeout(() => {
+        dismissMessage(id);
+        scheduledMessages.current.delete(id);
+      }, 5000);
```

Comment:
I have been struggling in the tests to mock calls to `setTimeout`. I have tried using `sinon.useFakeTimers()` (like in `VideoPlayerApp-test`) as well as `sinon.stub(window, 'setTimeout').callsArg(0)`, and for some reason the callback was never invoked.

I have finally decided that having a test seam prop was simple enough.

Reply:
A test seam is fine with me.